planner/core/optimizer.go

```diff
@@ -148,6 +148,37 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	plan = eliminatePhysicalProjection(plan)
 	plan = injectExtraProjection(plan)
 	plan = eliminateUnionScanAndLock(sctx, plan)
+	plan = enableParallelApply(sctx, plan)
+	return plan
+}
+
+func enableParallelApply(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
```

Comment:
It is hard to do that.
Because of limitation 3, we have to iterate the `Plan` from its root node, but if we move this check to `buildApply(v *plannercore.PhysicalApply)`, we cannot know if the plan `v` is in the inner side of another `Apply`.
Besides, I think keeping this logic in the optimization phase is more rational and easy-to-understand~